src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,853 @@
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_string(m, NFTA_EXPR_NAME, name);
+        if (r < 0)
+                return r;
+
+        return sd_netlink_message_open_container_union(m, NFTA_EXPR_DATA, name);
+}
+
+static int nfnl_add_expr_fib(sd_netlink_message *m, uint32_t nft_fib_flags,
+                             enum nft_fib_result result,
+                             enum nft_registers dreg) {
+        int r;
+
+        r = nfnl_add_open_expr_container(m, "fib");
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_u32(m, NFTA_FIB_FLAGS, htobe32(nft_fib_flags));
+        if (r < 0)
+                return r;
+        r = sd_netlink_message_append_u32(m, NFTA_FIB_RESULT, htobe32(result));
+        if (r < 0)
+                return r;
+        r = sd_netlink_message_append_u32(m, NFTA_FIB_DREG, htobe32(dreg));
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_close_container(m); /* NFTA_EXPR_DATA */
+        if (r < 0)
+                return r;
+
+        return sd_netlink_message_close_container(m); /* NFTA_LIST_ELEM */
+}
+
+static int nfnl_add_expr_meta(sd_netlink_message *m, enum nft_meta_keys key,
+                              enum nft_registers dreg) {
+        int r;
+
+        r = nfnl_add_open_expr_container(m, "meta");
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_u32(m, NFTA_META_KEY, htobe32(key));
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_u32(m, NFTA_META_DREG, htobe32(dreg));
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_close_container(m); /* NFTA_EXPR_DATA */
+        if (r < 0)
+                return r;
+
+        return sd_netlink_message_close_container(m); /* NFTA_LIST_ELEM */
+}
+
+static int nfnl_add_expr_payload(sd_netlink_message *m, enum nft_payload_bases pb,
+                                 uint32_t offset, uint32_t len, enum nft_registers dreg) {
+        int r;
+
+        r = nfnl_add_open_expr_container(m, "payload");
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_u32(m, NFTA_PAYLOAD_DREG, htobe32(dreg));
+        if (r < 0)
+                return r;
+        r = sd_netlink_message_append_u32(m, NFTA_PAYLOAD_BASE, htobe32(pb));
+        if (r < 0)
+                return r;
+        r = sd_netlink_message_append_u32(m, NFTA_PAYLOAD_OFFSET, htobe32(offset));
+        if (r < 0)
+                return r;
+        r = sd_netlink_message_append_u32(m, NFTA_PAYLOAD_LEN, htobe32(len));
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_close_container(m); /* NFTA_EXPR_DATA */
+        if (r < 0)
+                return r;
+        return sd_netlink_message_close_container(m); /* NFTA_LIST_ELEM */
+}
+
+static int __nfnl_add_expr_lookup_set(sd_netlink_message *m, const char *set_name,
```

Comment:
hmm, double udnerscores are actually reserved by the compiler/C language. kernel uses them anyway, but let's avoid that here, and use single underscores at most at most.

That said, can't we find a better name for this, i.e. using human understandable words to discern it from nfnl_add_expr_lookup_set() that lacks the underscores?

Reply:
Yes, ugly habit of mine. I've renamed the functions and dropped the "__" prefix.